lib/_http_client.js

```diff
@@ -801,13 +801,15 @@ function onSocketNT(req, socket, err) {
       req.emit('close');
     }
 
-    if (!err && req.agent) {
-      socket?.emit('free');
-    } else if (socket) {
-      finished(socket.destroy(err || req[kError]), (er) => {
-        _destroy(req, er || err);
-      });
-      return;
+    if (socket) {
+      if (!err && req.agent && !socket.destroyed) {
+        socket.emit('free');
+      } else {
+        finished(socket.destroy(err || req[kError]), (er) => {
```

Comment:
The `'error'` listener added by `finished()` must be removed if the connection is upgraded.